kratos/utilities/data_type_traits.h

```diff
@@ -0,0 +1,642 @@
+        }
+    }
+
+    inline static PrimitiveType * GetContiguousData(ContainerType& rValue)
+    {
+        if constexpr(HasContiguousPrimitiveData) {
+            return rValue.data();
+        } else {
+            static_assert(!std::is_same_v<TDataType, TDataType>, "This should be only called if the rValue is contiguous only.");
+        }
+    }
+
+    inline static void FillToContiguousData(
+        PrimitiveType* pContiguousDataBegin,
+        const ContainerType& rContainer)
+    {
+        const auto stride = DataTypeTraits<ValueType>::Size(rContainer[0]);
+        for (unsigned int i = 0; i < Dimension; ++i) {
+            DataTypeTraits<ValueType>::FillToContiguousData(pContiguousDataBegin + i * stride, rContainer[i]);
+        }
+    }
+
+    inline static void FillFromContiguousData(
+        ContainerType& rContainer,
+        PrimitiveType const * pContiguousDataBegin)
+    {
+        const auto stride = DataTypeTraits<ValueType>::Size(rContainer[0]);
+        for (unsigned int i = 0; i < Dimension; ++i) {
+            DataTypeTraits<ValueType>::FillFromContiguousData(rContainer[i], pContiguousDataBegin + i * stride);
+        }
+    }
+
+    ///@}
+};
+
+template<class TDataType> class DataTypeTraits<DenseVector<TDataType>>
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ContainerType = DenseVector<TDataType>;
+
+    using ValueType = TDataType;
+
+    using PrimitiveType = typename DataTypeTraits<ValueType>::PrimitiveType;
+
+    static constexpr bool HasContiguousPrimitiveData = std::is_same_v<PrimitiveType, ValueType>;
+
+    static constexpr bool HasDynamicMemoryAllocation = true;
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    static inline unsigned int Size(const ContainerType& rValue)
+    {
+        return (rValue.size() == 0 ? 0 : rValue.size() * DataTypeTraits<ValueType>::Size(rValue[0]));
+    }
+
+    static inline std::vector<unsigned int> Shape(const ContainerType& rValue)
+    {
+        std::vector<unsigned int> shape;
+        if (rValue.size() == 0) {
+            shape = DataTypeTraits<ValueType>::Shape(ValueType{});
+        } else {
+            shape = DataTypeTraits<ValueType>::Shape(rValue[0]);
+        }
+        shape.insert(shape.begin(), rValue.size());
+        return shape;
+    }
+
+    static inline bool Reshape(
+        ContainerType& rContainer,
+        const std::vector<unsigned int>& rShape)
+    {
+        KRATOS_ERROR_IF_NOT(rShape.size() >= 1) << "Invalid shape given for DenseVector data type.";
+
+        bool is_reshaped = false;
+
+        if (rContainer.size() != rShape[0]) {
+            rContainer.resize(rShape[0], false);
+            is_reshaped = true;
+        }
+
+        if constexpr(DataTypeTraits<ValueType>::HasDynamicMemoryAllocation) {
+            std::vector<unsigned int> sub_data_type_shape(rShape.size() - 1);
+            std::copy(rShape.begin() + 1, rShape.end(), sub_data_type_shape.begin());
```

Comment:
this is why I suggest implementing stuff in terms of iterators. You'd save loads of unnecessary allocations and copies here.

Reply:
done :)